### PR TITLE
feat: add run_events_pubsub

### PR DESCRIPTION
**Pull request by grant**
Adds an Events sample using the CloudEvents SDK for Go for Pub/Sub events.

### CODE REVIEW

**Comment by grant** [comment]
Looks like the error is in the SDK:

```
/go/pkg/mod/github.com/cloudevents/sdk-go/v2@v2.0.0/types/value.go:306:27: reflect.ValueOf(v).IsZero undefined (type reflect.Value has no field or method IsZero)
note: module requires Go 1.13
```

File: run/events-pubsub/main.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+	handler = http.DefaultServeMux
+)
+
+func main() {
+	ctx := context.Background()
+	p, err := cloudevents.NewHTTP()
+	if err != nil {
+		log.Fatal(err)
+	}
+	handleFn, err := cloudevents.NewHTTPReceiveHandler(ctx, p, HelloPubSub)
```
[the remark this answers is not in the document]
Comment (grant) [reply run/events-pubsub/main.go:41]: Thanks.

File: run/events-pubsub/main.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+)
+
+var (
+	handler = http.DefaultServeMux
+)
+
+func main() {
+	ctx := context.Background()
+	p, err := cloudevents.NewHTTP()
```
[the remark this answers is not in the document]
Comment (grant) [reply run/events-pubsub/main.go:37]: Added comment.

File: run/events-pubsub/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (grant) [reply run/events-pubsub/main.go:41]: Good catch. Fixed.

File: run/events-pubsub/main.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+// [START run_events_pubsub_handler]
+
+// PubSubMessage is the payload of the Pub/Sub message.
+type PubSubMessage struct {
+	Data []byte `json:"data,omitempty"`
+	ID   string `json:"id"`
+}
+
+// PubSub is the whole payload of a Pub/Sub event.
+type PubSub struct {
```
[the remark this answers is not in the document]
Comment (grant) [reply run/events-pubsub/main.go:71]: Correct. We don't have external bindings yet.

This is what we do in other samples.

File: run/events-pubsub/main.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+// HelloPubSub receives and processes a Pub/Sub CloudEvent.
+func HelloPubSub(ctx context.Context, event cloudevents.Event) string {
+	// Try to decode the request body into the struct.
+	var m PubSub
+	err := event.DataAs(&m)
+	if err != nil {
+		// Error parsing CloudEvent
+		log.Printf("Could not read CloudEvent: %v", err)
```
[the remark this answers is not in the document]
Comment (grant) [reply run/events-pubsub/main.go]: Thanks. Applied.

File: run/events-pubsub/main.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+		return ""
+	}
+
+	// Print and return the data from the CloudEvent
+	s := fmt.Sprintf("Hello, %s! ID: %s", string(m.Message.Data), event.ID())
+	log.Print(s)
+	return s
```
[the remark this answers is not in the document]
Comment (grant) [reply run/events-pubsub/main.go:90]: This Go function accepts a CloudEvent and returns a string.
The return value is useful for tests but could be used like any other function with input and output.

I clarified a bit in the comment above the statements.

File: run/events-pubsub/main_test.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+		DataEncoded: data,
+	}
+	if ce == nil {
+		t.Log("aweg")
+	}
+
+	s := HelloPubSub(context.Background(), *ce)
+	if s != "Hello, foo! ID: 321-CBA" {
+		t.Error("Didn't get CloudEvent message.")
+	}
```
[the remark this answers is not in the document]
Comment (grant) [reply run/events-pubsub/main_test.go]: Applied.

**Comment by grant** [comment]
Thanks for the review @grayside. PTAL.

File: run/events-pubsub/main.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+// HelloPubSub receives and processes a Pub/Sub CloudEvent.
+func HelloPubSub(ctx context.Context, event cloudevents.Event) string {
+	// Try to decode the request body into the struct.
+	var m PubSub
+	err := event.DataAs(&m)
+	if err != nil {
+		// Error parsing CloudEvent
+		log.Printf("event.DataAs: could not read CloudEvent: %v", err)
+		return ""
```
Comment (grayside) [inline run/events-pubsub/main.go:84]: I'm still unclear on the response semantics of returning a string: does an empty string imply a HTTP 400 Status Code? If not, how do we indicate an error here so the Pub/Sub subscription will retry?

Comment (grant) [reply run/events-pubsub/main.go]: I think the CloudEvent HTTP Receiver handler should handle that logic.
I changed the return value to be able to return the error as well.

File: run/events-pubsub/main.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+	s := ""
+	if err != nil {
+		// Error parsing CloudEvent
+		s = fmt.Sprintf("event.DataAs: could not read CloudEvent: %v", err)
+	} else {
+		// Print and return the data from the Pub/Sub CloudEvent.
+		s = fmt.Sprintf("Hello, %s! ID: %s", string(m.Message.Data), event.ID())
+	}
+	log.Printf(s)
+	return s, err
```
Comment (grayside) [inline run/events-pubsub/main.go:90]: My expectation is that if we return an error, the handler will be responsible for logging it. Given that behavior, I'd expect this code to look like:

```suggestion
	if err != nil {
		// Error parsing CloudEvent
		return "", fmt.Errorf("event.DataAs: could not read CloudEvent: %v", err)
	}
	// Print and return the data from the Pub/Sub CloudEvent.
	s := fmt.Sprintf("Hello, %s! ID: %s", string(m.Message.Data), event.ID())
	log.Printf(s)
	return s, nil
```

Comment (grant) [reply run/events-pubsub/main.go]: Yes. That is more idiomatic.

I'm usually in the headspace where I don't like multiple returns in most languages, but this is far more idiomatic in Go.

**Comment by grayside** [comment]
Test failure: Not sure why Go 1.11 passed, but Go 1.12 is not compatible with the Events SDK. If the SDK isn't adjusted for backwards compatibility, this is blocked by #1224.

The current implement plans for that issue are skipping tests for a module in an earlier runtime than the module requires. I can look into a new iteration on solving that problem but odds are it will not land until July.

Extract from the build log:
```
/go/pkg/mod/github.com/cloudevents/sdk-go/v2@v2.0.0/types/value.go:306:27: reflect.ValueOf(v).IsZero undefined (type reflect.Value has no field or method IsZero)
note: module requires Go 1.13
FAIL	github.com/GoogleCloudPlatform/golang-samples/run/events-pubsub [build failed]
```

**Comment by grant** [comment]
t.Skip() because of Go failures

**Comment by grant** [comment]
https://github.com/cloudevents/sdk-go/issues/536

**Comment by broady** [comment]
Prefer undescores in package paths. Not written anywhere sorry but it's implied:
https://github.com/GoogleCloudPlatform/golang-samples/blob/master/CONTRIBUTING.md#sample-package-name-file-name-and-directory

Commit messages should follow the Go project style (e.g. functions: add gophers codelab).
https://github.com/GoogleCloudPlatform/golang-samples/blob/master/CONTRIBUTING.md

**Comment by grant** [comment]
> Prefer undescores in package paths. Not written anywhere sorry but it's implied:
> https://github.com/GoogleCloudPlatform/golang-samples/blob/master/CONTRIBUTING.md#sample-package-name-file-name-and-directory

Done. Although almost all other package paths for `run` don't do this (can we add a lint rule or something?)

> Commit messages should follow the Go project style (e.g. functions: add gophers codelab).
> https://github.com/GoogleCloudPlatform/golang-samples/blob/master/CONTRIBUTING.md

Can modify with the final squash commit.

**Comment by grant** [comment]
Hi @grayside, we talked about skipping this test in the meantime due to the Go CE SDK behavior.

I noticed you added `status: blocked`, however I think we're not blocked if we have verified the test in the above CI run and agreed to `t.Skip` right?

**Comment by grant** [comment]
I guess we can't just skip because the error is a build error.

Can we somehow skip the build for Go 1.12?

**Comment by grayside** [comment]
Ah, too true. There's not a graceful way to handle this. Kokoro testing is controlled by [system_tests.sh](https://github.com/GoogleCloudPlatform/golang-samples/blob/master/testing/kokoro/system_tests.sh) and in order to exempt a sample from even building, we would need to modify the script to exclude this directory in a few places.

1. To catch all the PR test points, modify `$CHANGED_DIRS` around line 46
2. To avoid nightly testing, would need to modify the runTests() function to remove this directory, or perhaps in the loop that calls it around line 204.

@tbpg we are trying to find a way to merge this PR with skipped tests until I can revisit #1224 in July. Do you have any suggestions?

**Comment by tbpg** [comment]
Build tags might work here. @broady mentioned not being a fan, though.

**Comment by grant** [comment]
Will wait for @broady to recommend a solution.

**Comment by grant** [comment]
@tbpg @broady Is there an interim solution we can apply in the meantime? Can we have the CI skip this repo's folder?

My Go knowledge and repo knowledge doesn't know how to move forward with this PR.

**Comment by broady** [comment]
see #1552

**Comment by grant** [comment]
Thanks @broady!
I'm not authorized to merge, so please merge when you can.

**Pull request merged**